exporter/latency.go

```diff
@@ -30,9 +35,77 @@ func (e *Exporter) extractLatencyMetrics(ch chan<- prometheus.Metric, c redis.Co
 			var spikeLast, spikeDuration, max int64
 			if _, err := redis.Scan(latencyResult, &eventName, &spikeLast, &spikeDuration, &max); err == nil {
 				spikeDurationSeconds := float64(spikeDuration) / 1e3
-				e.registerConstMetricGauge(ch, "latency_spike_last", float64(spikeLast), eventName)
-				e.registerConstMetricGauge(ch, "latency_spike_duration_seconds", spikeDurationSeconds, eventName)
+				exporter.registerConstMetricGauge(outChan, "latency_spike_last", float64(spikeLast), eventName)
+				exporter.registerConstMetricGauge(outChan, "latency_spike_duration_seconds", spikeDurationSeconds, eventName)
 			}
 		}
 	}
 }
+
+func extractLatencyHistogramMetrics(exporter *Exporter, outChan chan<- prometheus.Metric, redisConn redis.Conn) {
+	reply, err := redis.Values(doRedisCmd(redisConn, "LATENCY", "HISTOGRAM"))
+	if err != nil {
+		logHistogramErrOnce.Do(func() {
+			log.Errorf("WARNING, LOGGED ONCE ONLY: cmd LATENCY HISTOGRAM, err: %s", err)
+		})
+		log.Debugf("cmd LATENCY HISTOGRAM, err: %s", err)
+		return
+	}
+
+	type cmdDetails struct {
+		cmd     string
+		details []interface{}
+	}
+
+	cmdCount := len(reply) / 2
+	details := make([]cmdDetails, cmdCount)
+
+	for i := 0; i < cmdCount; i++ {
+		cmd := string(reply[i*2].([]byte))
+		det, _ := redis.Values(reply[i*2+1], nil)
+
+		details[i].cmd = cmd
+		details[i].details = det
+	}
+
+	for _, detail := range details {
+		var totalCalls uint64
+		var bucketInfo []uint64
+
+		buckets := make(map[float64]uint64)
+
+		if _, err := redis.Scan(detail.details, nil, &totalCalls, nil, &bucketInfo); err != nil {
+			break
+		}
+
+		for j := 0; j < len(bucketInfo); j += 2 {
+			usec := float64(bucketInfo[j])
+			count := bucketInfo[j+1]
+			buckets[usec] = count
+		}
+
+		sumEstimated := estimateSum(buckets)
+
+		labelValues := []string{"cmd"}
+		exporter.registerConstHistogram(outChan, "latency_usec", labelValues, totalCalls, sumEstimated, buckets, detail.cmd)
+	}
+}
+
+func estimateSum(buckets map[float64]uint64) float64 {
+	// This should ideally be the total time as reported by info commandstats.
+	// Estimated here instead of querying redis again.
+
+	sum := .0
+	counted := uint64(0)
+
+	for usec, count := range buckets {
+		bucketCount := count - counted
```

Comment:
I'm not sure I get this logic.

my understanding is, you're trying to go through the buckets and estimage the avg for each bucket (`bucketAverage`) and then multiply by the times of events in that bucket to get the total latency for that bucket and then sum all the buckets, right?

I'm not sure I understand bucketCount.
But there's a golang idiosyncrasy and that is looping over maps is always in random order of the keys so you don't have any guarantee in which order you'll look at the buckets so I think your "counted=count" logic won't work.

But look at the docs: https://redis.io/commands/latency-histogram/
the buckets increase by power of 2 (1, 2, 4, 8, ...) so you could use that information to look through the buckets (and have better estimates.

Or you could toss this all and use the cmd stats as retrieved earlier (which i think makes more sense)